posts/2019-07-04-Rust-1.36.0.md

```diff
@@ -0,0 +1,164 @@
+[`mem::uninitialized`]: https://doc.rust-lang.org/std/mem/fn.uninitialized.html
+[gankro-blog]: https://gankro.github.io/blah/initialize-me-maybe/
+[pr-60445]: https://github.com/rust-lang/rust/pull/60445
+
+In previous releases of Rust, the [`mem::uninitialized`] has allowed you bypass Rust's initialization checks by pretending that you've initialized a value at type `T` without doing anything. One of the main uses of this function has been to lazily allocate arrays.
+
+However, [`mem::uninitialized`] is an incredibly dangerous operation that essentially cannot be used correctly as the Rust compiler assumes that values are properly initialized. For example, calling `mem::uninitialized::<bool>()` causes *instantaneous __undefined behavior__* as the random bits in memory could be something other than `0` and `1`.
+
+To remedy this situation, in Rust 1.36.0, the type [`MaybeUninit<T>`] has been [stabilized][pr-60445]. The Rust compiler will understand that it should not assume that a [`MaybeUninit<T>`] is a properly initialized `T`. Therefore, you can do gradual initialization more safely and eventually use `.assume_init()` once you are certain that `maybe_t: MaybeUninit<T>` contains an initialized `T`.
+
+As [`MaybeUninit<T>`] is the safer alternative, starting with Rust 1.38, the function [`mem::uninitialized`] will be deprecated.
+
+To read more about uninitialized memory, [`mem::uninitialized`], and [`MaybeUninit<T>`], read [Alexis Beingessner's blog post][gankro-blog] about the subject. The standard library also contains extensive documentation about [`MaybeUninit<T>`].
+
+### NLL on Rust 2015
+
+[nll-2018]: https://blog.rust-lang.org/2018/12/06/Rust-1.31-and-rust-2018.html#non-lexical-lifetimes
+[soundness]: https://en.wikipedia.org/wiki/Soundness
+[felix-blog]: http://blog.pnkfx.org/blog/2019/06/26/breaking-news-non-lexical-lifetimes-arrives-for-everyone/
+[crater-nll]: https://github.com/rust-lang/rust/issues/60680#issuecomment-495089654
+
+[Previously in Rust 1.31.0][nll-2018] we told you about NLL (Non-Lexical Lifetimes), an improvement to the language that makes the borrow checker smarter and more user friendly. For example, you may now write:
+
+```rust
+fn main() {
+    let mut x = 5;
+    let y = &x;
+    let z = &mut x; // This was not allowed before 1.31.0.
+}
+```
+
+In 1.31.0 NLL was stabilized for Rust 2018 only with a promise that we would back port it to Rust 2015 as well. With Rust 1.36.0, we are happy to announce that we have done so! NLL is now available in Rust 2015 as well.
+
+With NLL on both editions, we are closer to removing the old borrow checker. However, the old borrow checker unfortunately accepted some [unsound][soundness] code it should not have. As a result, NLL is currently in a "migration mode" wherein we will emit warnings instead of errors if the NLL borrow checker rejects code the old AST borrow checker would accept. Please see [this list][crater-nll] to see if one of your crates are affected.
```

Comment:
Private code may be affected, and non-owners may be indirectly affected.

```suggestion
With NLL on both editions, we are closer to removing the old borrow checker. However, the old borrow checker unfortunately accepted some [unsound][soundness] code it should not have. As a result, NLL is currently in a "migration mode" wherein we will emit warnings instead of errors if the NLL borrow checker rejects code the old AST borrow checker would accept. Please see [this list][crater-nll] of public crates that are affected.
```